msteams-platform/concepts/bots/bots-files.md

```diff
@@ -0,0 +1,152 @@
+
+## Overview
+
+One way of doing this is using the Microsoft Graph APIs for OneDrive and SharePoint, [as documented here](https://docs.microsoft.com/en-us/onedrive/developer/rest-api/). This method requires obtaining ongoing access to the user's OneDrive folder through standard OAuth2 authorization flow. This documentation describes a simple alternative mechanism if you only need to send file content as a result of direct user interaction, e.g. sending a message. This API is provided as part of the Microsoft Teams Bot Platform.
+
+## Receiving files in 1:1 chat
+
+When a user sends a file to your bot, the file is first uploaded to the user's OneDrive for Business storage. Your bot will then receive a message activity notifying you of the user upload. The activity will contain file metadata, such as its name and the content URL, and the content URL's expiry time (e.g. 24 hours). You can directly read from this URL to fetch its binary content.
+
+### Message activity with file attachment example
+
+```json
+{
+  "attachments": [{
+    "contentType": "application/vnd.microsoft.teams.file.download.info",
+    "contentUrl": "https://contoso.sharepoint.com/personal/johnadams_contoso_com/Documents/Applications/file_example.txt",
+    "name": "file_example.txt",
+    "content": {
+      "downloadUrl" : "https://download.link",
+      "uniqueId": "1150D938-8870-4044-9F2C-5BBDEBA70C9D",
+      "fileType": "txt",
+      "etag": "123"
+    }
+  }]
+}
+```
+
+As a best practice, you should acknowledge the file upload by sending back a message to the user.
+
+## Uploading files to 1:1 chat
+
+Uploading a file to a user involves the following steps:
+
+1. Send a message to the user requesting permission to write the file. This message must contain a `FileConsentCard` attachment with the name of the file to be uploaded.
+2. If the user accepts the file download, your bot will receive an Invoke activity with a location URL.
+3. To transfer the file, your bot performs an HTTP POST directly into the provided location URL.
+4. Optionally, you can perform a message update on the original message, removing the consent card if you do not want to allow the user to accept further uploads of the same file.
+
+### Message requesting permission to upload
+
+This message contains a simple attachment object requesting user permission to upload the file.
+
+```json
+{
+  "attachments": [{
+    "contentType": "application/vnd.microsoft.teams.card.file.consent",
+    "name": "file_example.txt",
+    "content": {
+      "description": "<Purpose of the file, e.g. this is your monthly expense report>",
+      "sizeInBytes": 1029393,
```

Comment:
Let's describe the purpose of this sizeInBytes. Does this have to be exact, or is an approximate value ok? Seems unreasonable to ask the bot to produce the file (to calculate the exact size) if the user might deny consent or ignore the request.

Reply:
It's just an estimate.